README.md
# harbor-compose

A tool for defining and running multi-container Docker applications on Harbor.

[![CircleCI](https://circleci.com/gh/turnerlabs/harbor-compose/tree/master.svg?style=shield)](https://circleci.com/gh/turnerlabs/harbor-compose/tree/master)  

With Harbor Compose, you start with a standard [Docker Compose](https://docs.docker.com/compose/) file to configure your application’s services. You then add a Harbor Compose file to configure Harbor-specific settings.  Then, using a single command, you create and start all the services from your configuration.

Using Harbor Compose is basically a four-step process.

1. Define your app’s environment with a `Dockerfile` so it can be reproduced anywhere.

2. Define the services that make up your app in `docker-compose.yml` so they can be run together in an isolated environment.  You can use the standard Docker Compose commands (like `docker-compose build`, `docker-compose push`, `docker-compose up`, etc.) to build/run/test your Docker app locally.

3. When you're ready to launch your Docker app on Harbor, you define the Harbor-specific parameters in a [`harbor-compose.yml`](compose-reference.md) file.

4. Run `harbor-compose up` and Harbor Compose will start and run your entire app on a managed barge.


Just like `docker-compose`, `harbor-compose` has similar commands for managing the life cycle of your app on Harbor:

- Start and stop services
- Upload environment variables and scale replicas
- View the status of running services
- Stream the log output of running services
- Trigger an image deployment from a public build system (like Circle CI, Travis CI, etc.)


A simple [`docker-compose.yml`](https://docs.docker.com/compose/compose-file/compose-file-v2/) might look like this:

```yaml
version: "2"
services:
  web-app:
    image: quay.io/turner/my-web-app:1.0.0
    ports:
      - "80:5000"
    environment:
      HEALTHCHECK: /hc
      PORT: 5000
```

A [`harbor-compose.yml`](compose-reference.md) might look like this:

```yaml
version: "1"
shipments:
  my-web-app:
    env: dev
    barge: digital-sandbox
    containers:
      - web-app    
    replicas: 2
    group: mss
    property: turner
    project: turner
    product: turner
```

Then to start your application...

```
$ harbor-compose up
```

Access your app logs...

```
$ harbor-compose logs
```

Get the status of your shipments...

```
$ harbor-compose ps
```

To stop your application, remove all running containers and delete your load balancer...

```
$ harbor-compose down
```

### Compose file reference

See the [full harbor-compose.yml reference](compose-reference.md) along with which [docker-compose.yml](https://docs.docker.com/compose/) properties are supported by Harbor Compose.


#### Getting Started

There are currently two installation options for Harbor Compose.

1) Download the latest binary from [Github releases](https://github.com/turnerlabs/harbor-compose/releases).

```
curl -s get-cli.harbor.turnerlabs.io | sh
```

2) Run as a docker container

```
docker run -it --rm -v `pwd`:/work quay.io/turner/harbor-compose up
```

- or if you want to reuse your session:

```
docker run -it —rm -v `pwd`:/work -v ${HOME}/.harbor:/root/.harbor quay.io/turner/harbor-compose up
```


To get started with an existing shipment, you can run the following to generate [`docker-compose.yml`](https://docs.docker.com/compose/compose-file/compose-file-v2/) and [`harbor-compose.yml`](compose-reference.md) files, by specifying the shipment name and environment as args.  Note that you will be prompted to login if you don't already have a token or if your token has expired.  For example:

```
$ harbor-compose generate my-shipment dev
```

To create new shipments, you can use the `init` command to generate new [`docker-compose.yml`](https://docs.docker.com/compose/compose-file/compose-file-v2/) and [`harbor-compose.yml`](compose-reference.md) files.  `init` will ask you questions to build your compose files.  Note that you use the `--yes` flag to accept defaults and generate one quickly.

```
$ harbor-compose init
```

This will output the files in the current directory.  You can then run a bunch of useful commands, for example...

Run your shipment locally in Docker (practically identically to how it runs in Harbor)...

```
$ docker-compose up
```

Push your images to a registry.

```
$ docker-compose push
```

Scale your shipment by changing the replicas in `harbor-compose.yml`, or change your environment variables and re-deploy, or deploy a new image, etc....

```
$ harbor-compose up
```

Get the status of your shipment(s) using the `ps` command.  With this command you can see the status of each container replica, when it started and the last known state.  For example:

```
$ harbor-compose ps

SHIPMENT:      mss-poc-multi-container   
ENVIRONMENT:   dev                       
STATUS:        Running                   
CONTAINERS:    2                         
REPLICAS:      2

ID        IMAGE                                                        STATUS    STARTED      RESTARTS   LAST STATE              
ab97cef   quay.io/turner/mss-poc-multi-container:1.0.0    running   1 week ago   1          terminated 1 week ago   
873a390   quay.io/turner/mss-poc-multi-container2:1.0.0   running   1 week ago   1          terminated 1 week ago   
73fad42   quay.io/turner/mss-poc-multi-container:1.0.0    running   5 days ago   2          terminated 5 days ago   
db93650   quay.io/turner/mss-poc-multi-container2:1.0.0   running   5 days ago   2          terminated 5 days ago   
```

You can also manage multiple shipments using Harbor Compose by listing them in your harbor-compose.yml file.  This is particularly useful if you have a web/worker, or microservices type application where each shipment can be scaled independently.


#### Authentication

Some commands (`up`, `down`, `generate`) require authentication and will automatically prompt you for your credentials.  A temporary (6 hours) authentication token is stored on your machine so that you don't have to login when running each command.  If you want to logout and remove the authentication token, you can run the `logout` command.  You can also explicitly login by running the `login` command.


#### CI/CD

See the [CI/CD doc](cicd.md).


CHANGELOG.md
## 0.18.1 (2018-10-08)

Features:

- v0.6.1 of the migration template ([#196](https://github.com/turnerlabs/harbor-compose/pull/196))


## 0.18.0 (2018-09-20)

Features:

- Adds `migrate` command


## 0.17.0 (2018-02-07)

Features:

- Add support for `roles` to `init` command ([#174](https://github.com/turnerlabs/harbor-compose/issues/174))

- adds `loadbalancer` and `iam_role` to terraform output ([#177](https://github.com/turnerlabs/harbor-compose/issues/177))

- Add an `events` command to surface the k8s events in the helmit api ([#118](https://github.com/turnerlabs/harbor-compose/issues/118))

- adds fields to telemetry payload ([#173](https://github.com/turnerlabs/harbor-compose/issues/173))


Bug fixes:

- fixes file permissions with circle ci build provider ([#178](https://github.com/turnerlabs/harbor-compose/issues/178))



## 0.16.0 (2018-01-17)

Features:

- Adds new `env` command (ls, push, pull) ([#165](https://github.com/turnerlabs/harbor-compose/issues/165), [#137](https://github.com/turnerlabs/harbor-compose/issues/137), [#121](https://github.com/turnerlabs/harbor-compose/issues/121))

- `ps` and `logs` commands now work without yaml files ([#163](https://github.com/turnerlabs/harbor-compose/issues/163), [#150](https://github.com/turnerlabs/harbor-compose/issues/150))

- Adds `buildtoken env` command ([#164](https://github.com/turnerlabs/harbor-compose/issues/164))

- commands that output terraform should omit duplicate config in harbor-compose.yml ([#167](https://github.com/turnerlabs/harbor-compose/issues/167))

- telemetry integration ([#151](https://github.com/turnerlabs/harbor-compose/issues/151))


Bug fixes:

- `generate` should write log shipping env vars to harbor-compose.yml ([#141](https://github.com/turnerlabs/harbor-compose/issues/141))

- The `up` command should require and validate that env vars are not empty ([#138](https://github.com/turnerlabs/harbor-compose/issues/138))

- The `generate` command should properly escape env vars with $ characters ([#153](https://github.com/turnerlabs/harbor-compose/issues/153))

- healthcheckIntervalSeconds must be > healthcheckTimeoutSeconds ([#140](https://github.com/turnerlabs/harbor-compose/issues/140))


## 0.15.0 (2017-11-10)

Features:

- Add support for generating terraform source files ([#142](https://github.com/turnerlabs/harbor-compose/issues/142))

- Generate a terraform main.tf from 'init' command ([#143](https://github.com/turnerlabs/harbor-compose/issues/143))

- friendly dns install ([#149](https://github.com/turnerlabs/harbor-compose/issues/149))


## 0.14.0 (2017-9-25)

Features:

- Support for `hidden` environment variables ([#19](https://github.com/turnerlabs/harbor-compose/issues/19))

- Output a `hidden.env` in the `init` command ([#125](https://github.com/turnerlabs/harbor-compose/issues/125))

- More efficient environment variable updates in `up` ([#126](https://github.com/turnerlabs/harbor-compose/issues/126))

- Support for `enableMonitoring` ([#79](https://github.com/turnerlabs/harbor-compose/issues/79))

- Support for `healthcheckIntervalSeconds` ([#129](https://github.com/turnerlabs/harbor-compose/issues/129))

- Support for `healthcheckTimeoutSeconds` ([#78](https://github.com/turnerlabs/harbor-compose/issues/78))

- Adds `barge` to `ps` command output ([#134](https://github.com/turnerlabs/harbor-compose/issues/134))


## 0.13.1 (2017-8-31)

Bug fixes:

- Codeship fix ([#122](https://github.com/turnerlabs/harbor-compose/issues/122))

- Restart fix ([#123](https://github.com/turnerlabs/harbor-compose/issues/123))


## 0.13.0 (2017-8-28)

Features:

- Adds Codeship build provider (`generate --build-provider codeship`) ([#101](https://github.com/turnerlabs/harbor-compose/issues/101))

- Adds `restart` command ([#109](https://github.com/turnerlabs/harbor-compose/issues/109))

- Asterisks are now displayed when entering password for login. ([#113](https://github.com/turnerlabs/harbor-compose/issues/113))

- Adds richer validation/feedback for `up` command ([#111](https://github.com/turnerlabs/harbor-compose/issues/111))

- Adds environment name validation for `up` command ([#110](https://github.com/turnerlabs/harbor-compose/issues/110))


Bug fixes:

- The `ps` command doesn't work when using expanded docker compose build context ([#112](https://github.com/turnerlabs/harbor-compose/issues/112))


## 0.12.1 (2017-7-24)

Bug fixes:

- generate --build-provider circleciv2 now uses `docker 17.06.0-ce` (to support multi-stage build)


## 0.12.0 (2017-7-12)

Features:

- Harbor Compose now scaffolds out `Circle CI v2 config files` that enable `CI/CD pipelines` into Harbor (with more providers to come). ([#95](https://github.com/turnerlabs/harbor-compose/issues/95))

- Added a new `buildtoken` command (and sub commands) for managing Harbor build tokens ([#100](https://github.com/turnerlabs/harbor-compose/issues/100))

- The `init` command now outputs `docker-compose.yml` files to make it easier to Dockerize/Harborize ([#35](https://github.com/turnerlabs/harbor-compose/issues/35))

- The `ps` command now includes service endpoint(s) ([#94](https://github.com/turnerlabs/harbor-compose/issues/94))

- All commands should return a useful exit code should ([#96](https://github.com/turnerlabs/harbor-compose/issues/96))

- Added ability to configure harbor-compose via a config file ([#97](https://github.com/turnerlabs/harbor-compose/issues/97))


Bug fixes:

- The up command continues to run even if user is not authenticated ([#102](https://github.com/turnerlabs/harbor-compose/issues/102))


## 0.11.0 (2017-5-16)

Features:

- `generate --build-provider` feature along with `circleciv1` implementation ([#84](https://github.com/turnerlabs/harbor-compose/issues/84))

- Use of variable substitution in docker-compose.yml for things other than envvars ([#82](https://github.com/turnerlabs/harbor-compose/issues/82))

Bug fixes:


cmd/init.go
package cmd

import (
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"strconv"
	"time"

	dockerProject "github.com/docker/libcompose/project"
	"github.com/spf13/cobra"
)

var yesUseDefaults bool

// initCmd represents the init command
var initCmd = &cobra.Command{
	Use:   "init",
	Short: "Interactively create docker-compose.yml, harbor-compose.yml, and main.tf files",
	Long: `This will ask you a bunch of questions, and then write a bunch of useful files for you.

docker-compose.yml = used to run locally in docker
hidden.env = used to store hidden environment variables that works in docker and in harbor
harbor-compose.yml = used to run remotely in harbor
main.tf = used to manage harbor infrastructure

If you invoke init with -y or --yes it will use only defaults and not prompt you for any options.`,
	Example: `harbor-compose init
harbor-compose init --yes
harbor-compose init -y`,
	Run:    initHarborCompose,
	PreRun: preRunHook,
}

func init() {
	initCmd.PersistentFlags().BoolVarP(&yesUseDefaults, "yes", "y", false, "don't ask questions and use defaults")
	RootCmd.AddCommand(initCmd)
}

const (
	defaultBackendImage = "quay.io/turner/turner-defaultbackend:0.2.0"
)

func initHarborCompose(cmd *cobra.Command, args []string) {

	//write a gitignored/dockerignored hidden.env as a placeholder for users to add secrets
	//if it already exists, prompt to overwrite
	//note that this file needs to be written before reference is added/parsed in docker-compose.yml
	write := false
	if _, err := os.Stat(hiddenEnvFileName); err == nil {
		fmt.Print(hiddenEnvFileName + " already exists. Overwrite? ")
		write = askForConfirmation()
	} else { //not exists
		write = true
	}
	if write {
		sampleContents := "#FOO=bar\n"
		err := ioutil.WriteFile(hiddenEnvFileName, []byte(sampleContents), 0644)
		check(err)
		sensitiveFiles := []string{hiddenEnvFileName, ".terraform"}
		appendToFile(".gitignore", sensitiveFiles)
		appendToFile(".dockerignore", sensitiveFiles)
	}

	//docker-compose
	registry := "quay.io/turner"
	container := "mss-harbor-app"
	tag := "0.1.0"
	publicPort := "80"
	internalPort := "5000"
	healthCheck := "/health"

	//harbor-compose
	name := "mss-harbor-app"
	env := "dev"
	barge := "digital-sandbox"
	replicas := "4"
	group := "mss"
	property := "turner"
	project := "turner"
	product := "turner"
	enableMonitoring := "true"
	hcTimeout := "1"
	hcInterval := "10"

	//generate a randomly generated human-readable name for defaults
	rand.Seed(time.Now().UTC().UnixNano())
	randomName := GetRandomName(0)

	//when -y is specified, use random shipment/container name
	if yesUseDefaults {
		name = randomName
		container = randomName
	}

	//if docker-compose.yml doesn't exist, then create one
	var dockerCompose DockerCompose
	if _, err := os.Stat(DockerComposeFile); err != nil {

		if !yesUseDefaults {
			registry = promptAndGetResponse("docker registry: (e.g., quay.io/turner) ", registry)
			container = promptAndGetResponse("docker container name: (e.g., mss-harbor-app) ", randomName)
			tag = promptAndGetResponse("version tag: (e.g., 0.1.0) ", tag)
			publicPort = promptAndGetResponse("public port: (e.g., 80) ", publicPort)
			internalPort = promptAndGetResponse("internal port: (e.g., 5000) ", internalPort)
			healthCheck = promptAndGetResponse("health check: (e.g., /health) ", healthCheck)
		}

		//create a DockerCompose object
		dockerCompose = DockerCompose{
			Version:  "2",
			Services: map[string]*DockerComposeService{},
		}

		image := fmt.Sprintf("%v/%v:%v", registry, container, tag)

		//swap default image with default backend image
		if image == "quay.io/turner/mss-harbor-app:0.1.0" || image == fmt.Sprintf("%v/%v:%v", registry, randomName, tag) {
			image = defaultBackendImage
		}

		//build a docker-compose object
		service := DockerComposeService{
			Build: ".",
			Image: image,
			Ports: []string{fmt.Sprintf("%v:%v", publicPort, internalPort)},
			Environment: map[string]string{
				"HEALTHCHECK": healthCheck,
			},
			EnvFile: []string{hiddenEnvFileName},
		}

		//add PORT env var if using default backend image
		if image == defaultBackendImage {
			service.Environment["PORT"] = internalPort
		}

		dockerCompose.Services[container] = &service

		//write docker-compose.yml
		SerializeDockerCompose(dockerCompose, DockerComposeFile)
	}

	intReplicas, _ := strconv.Atoi(replicas)
	monitoring, _ := strconv.ParseBool(enableMonitoring)
	healthcheckTimeoutSeconds, _ := strconv.Atoi(hcTimeout)
	healthcheckIntervalSeconds, _ := strconv.Atoi(hcInterval)
	var err error
	useRole := ""
	samlUser := ""

	//ask questions for harbor-compose.yml
	if !yesUseDefaults {
		name = promptAndGetResponse("shipment name: (e.g., mss-harbor-app) ", randomName)
		env = promptAndGetResponse("shipment environment: (dev, qa, prod, etc.) ", env)
		barge = promptAndGetResponse("barge: (digital-sandbox, ent-prod, corp-sandbox, corp-prod, news, nba) ", barge)
		useRole = promptAndGetResponse("would like you to use an AWS IAM Role? (y|n) ", "n")
		if useRole == "y" || useRole == "Y" {
			samlUser = promptAndGetResponse("which SAML user would you like to grant role access to? (e.g., aws-digital-sandbox-devops/First.Last@turner.com) ", "")
			if samlUser == "" {
				check(errors.New(messageSamlUserRequired))
			}
		}
		replicas = promptAndGetResponse("how many container instances: (e.g., 4) ", replicas)
		intReplicas, err = strconv.Atoi(replicas)
		if err != nil {
			check(errors.New(messageReplicasMustBeNumber))
		}
		group = promptAndGetResponse("group (mss, news, nba, ams, etc.): ", group)
		enableMonitoring = promptAndGetResponse("enable monitoring (true|false): ", enableMonitoring)
		monitoring, err = strconv.ParseBool(enableMonitoring)
		if err != nil {
			check(errors.New(messageEnableMonitoringTrueFalse))
		}
		hcTimeout = promptAndGetResponse("healthcheck timeout seconds (1): ", hcTimeout)
		healthcheckTimeoutSeconds, err = strconv.Atoi(hcTimeout)
		if err != nil {
			check(errors.New(messageTimeoutValidNumber))
		}
		hcInterval = promptAndGetResponse("healthcheck interval seconds (10): ", hcInterval)
		healthcheckIntervalSeconds, err = strconv.Atoi(hcInterval)
		if err != nil {
			check(errors.New(messageIntervalValidNumber))
		}
		//healthcheckIntervalSeconds must be > healthcheckTimeoutSeconds
		if !(healthcheckIntervalSeconds > healthcheckTimeoutSeconds) {
			check(errors.New(messageIntervalGreaterThanTimeout))
		}
		property = promptAndGetResponse("property (turner.com, cnn.com, etc.): ", property)
		project = promptAndGetResponse("project: ", project)
		product = promptAndGetResponse("product: ", product)
	}

	//create a harbor compose object
	harborCompose := HarborCompose{
		Shipments: make(map[string]ComposeShipment),
	}

	monitoring, err = strconv.ParseBool(enableMonitoring)
	if err != nil {
		check(errors.New(messageEnableMonitoringTrueFalse))
	}

	composeShipment := ComposeShipment{
		Env:                        env,
		Group:                      group,
		Property:                   property,
		Project:                    project,
		Product:                    product,
		EnableMonitoring:           &monitoring,
		HealthcheckTimeoutSeconds:  &healthcheckTimeoutSeconds,
		HealthcheckIntervalSeconds: &healthcheckIntervalSeconds,
	}

	composeShipment.Barge = barge
	composeShipment.Replicas = intReplicas

	//use existing docker-compose.yml to get containers
	dockerComposeProj := DeserializeDockerCompose(DockerComposeFile)
	proj := dockerComposeProj.(*dockerProject.Project)

	//add all docker services as containers
	for _, service := range proj.ServiceConfigs.Keys() {
		composeShipment.Containers = append(composeShipment.Containers, service)
	}

	//add single shipment to list
	harborCompose.Shipments[name] = composeShipment

	//transform compose yaml into a ShipmentEnvironment object
	shipmentEnvironment := transformComposeToShipmentEnvironment(name, composeShipment, dockerComposeProj)

	//generate a main.tf and write it to disk
	role := (useRole == "y" || useRole == "Y")
	generateAndWriteTerraformSource(&shipmentEnvironment, &harborCompose, false, role, samlUser)

	//add local dev role support to docker-compose.yml
	if role {
		addAwsRoleSupportToDockerComposeFile()
	}

	// remove duplicate properties (already in main.tf)
	harborCompose = minimalHarborCompose(harborCompose)

	//if harbor-compose.yml exists, ask to overwrite
	write = true
	if _, err := os.Stat(HarborComposeFile); err == nil {
		fmt.Print("harbor-compose.yml already exists. Overwrite? ")
		write = askForConfirmation()
	}
	if write {
		SerializeHarborCompose(harborCompose, HarborComposeFile)
	}

	fmt.Println("done")
	fmt.Println()
	fmt.Println("use terraform plan/apply to manage your infrastructure")
	fmt.Println("use docker-compose build/push, harbor-compose up/deploy to manage your application")
}

func promptAndGetResponse(question string, defaultResponse string) string {
	fmt.Print(question)
	var response string
	_, err := fmt.Scanln(&response)
	if err != nil && err.Error() != "unexpected newline" {
		log.Fatal(err)
	}
	if response == "" {
		response = defaultResponse
	}
	return response
}

func addAwsRoleSupportToDockerComposeFile() {
	roleYaml := getDockerComposeAwsRoleSupportYaml()

	//append to docker-compose.yml
	f, err := os.OpenFile(DockerComposeFile, os.O_APPEND|os.O_WRONLY, 0600)
	check(err)
	defer f.Close()
	_, err = f.WriteString(roleYaml)
	check(err)
}

func getDockerComposeAwsRoleSupportYaml() string {
	return `
# aws role support 

  role:
    image: quay.io/turner/ectou-metadata
    ports:
    - 9000:80
    environment:

      # the role you want your container to assume
      ROLE: arn:aws:iam::123456789:role/my-role

      # the local profile you want to use to assume the role
      AWS_PROFILE: ${AWS_PROFILE}

    volumes:
    - $HOME/.aws/credentials:/root/.aws/credentials:ro
    networks:
      default:
        ipv4_address: 169.254.169.254

networks:
  default:
    driver: bridge
    ipam:
     config:
     - subnet: 169.254.169.0/16
       gateway: 169.254.169.1
`
}
